toolchain/semantics/semantics_node.h

```diff
@@ -0,0 +1,129 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_SEMANTICS_SEMANTICS_NODE_H_
+#define CARBON_TOOLCHAIN_SEMANTICS_SEMANTICS_NODE_H_
+
+#include <cstdint>
+
+#include "common/ostream.h"
+#include "toolchain/semantics/semantics_node_kind.h"
+
+namespace Carbon {
+
+// Type-safe storage of Node IDs.
+struct SemanticsNodeId {
+  SemanticsNodeId() : id(-1) {}
+  explicit SemanticsNodeId(int32_t id) : id(id) {}
+  SemanticsNodeId(SemanticsNodeId const&) = default;
+  auto operator=(const SemanticsNodeId& other) -> SemanticsNodeId& = default;
+
+  void Print(llvm::raw_ostream& out) const { out << "%" << id; }
+
+  int32_t id;
+};
+
+// Type-safe storage of identifiers.
+struct SemanticsIdentifierId {
+  SemanticsIdentifierId() : id(-1) {}
+  explicit SemanticsIdentifierId(int32_t id) : id(id) {}
+
+  void Print(llvm::raw_ostream& out) const { out << "%s" << id; }
+
+  int32_t id;
+};
+
+// Type-safe storage of identifiers.
+struct SemanticsIntegerLiteralId {
+  SemanticsIntegerLiteralId() : id(-1) {}
+  explicit SemanticsIntegerLiteralId(int32_t id) : id(id) {}
+
+  void Print(llvm::raw_ostream& out) const { out << "%s" << id; }
+
+  int32_t id;
+};
+
+struct SemanticsTwoNodeIds {
+  SemanticsNodeId nodes[2];
+};
+
+union SemanticsNodeArgs {
+  SemanticsNodeArgs() {}
+  explicit SemanticsNodeArgs(SemanticsNodeId one_node) : one_node(one_node) {}
+  explicit SemanticsNodeArgs(SemanticsTwoNodeIds two_nodes)
+      : two_nodes(two_nodes) {}
+  explicit SemanticsNodeArgs(SemanticsIdentifierId identifier)
+      : identifier(identifier) {}
+  explicit SemanticsNodeArgs(SemanticsIntegerLiteralId integer_literal)
+      : integer_literal(integer_literal) {}
+
+  int no_args[0];
```

Comment:
Zero-length arrays aren't part of standard C++. If you want to have a named member for this case (eg, to make the code generated via your `.def` file work), consider using an empty struct instead.

Reply:
Done